### ISSUE TITLE
Create a flexible organizational model

### ISSUE THREAD

**Issue by jedelman8** [nautobot/nautobot#230]
### Proposed Functionality

The proposed functionality is to keep and maintain a default of an organizational structure as there is today with regions, sites, locations, etc. , but add the ability to make this more flexible.  This means users should be able to change the name of location types such as from sites to facilities and/or create abstract _containers_ to account for IDFs, MDFs, closets, buildings, campuses, sites, and regions (just as an example).  This should also account for contact information per abstract location too.  Using this as it will supercede #8 
By default, there should remain an opinionated org model.  Best case is, there are a handful of "org" templates to choose from to streamline the user experience, beyond just offering maximum flexibility for the users. 

### Use Case

Support different organization structures as they differ in the real world.

**Comment by sdktr** [comment]
Would a generic hierarchical 'Location' modal be a safer / more generic grouping? Instead of adding Buildings, Floors, Rooms as separate entities? Netbox 2.11 takes this route as well

**Comment by chadell** [comment]
# As P.D. - Plugin Developer
# I want ...
- to model a persona/contact that I could link to other other models, with the necessary information to identify and contact him/her

# So that...
- I could use this in multiple scenarios. For instance:
  - In the LifeCycle Plugin, there is no need to have a Contract POCs, simply linking a  `contract` to `contact` 
  - In a potential ToIP Plugin, the `ip_phone` could link to the `contact` 

The alternative to use Django `user` mixes to different ideas, as these `contacts` are not actual Django `users`, only model `personas`

# I know this is done when...
- I can represent `contacts` and `teams` (groups of contacts)
- I can create relationships to `contacts` from other models.

**Comment by bryanculver** [comment]
@nautobot/core is this a duplicate of #774 which might be a duplicate of #1052 and #1891?

**Comment by glennmatthews** [comment]
I read this issue (and especially the comments) as going beyond those issues into adding a generic Contact model that can be many-to-many associated to ~Sites~ Locations. Probably would be good to update the title and description to make that clearer.

**Comment by bryanculver** [comment]
OK I saw

> (1) introduce the concept of buildings such that one or more buildings can be added per site
> (2) consider creating a mechanism (using labels or some other means) to change the name of these organization objects.

Which to me was the Location model but @chadell's update constrains that. Since it doesn't look like we have an issue already open for a generic contact model (I could have swore we did) I can update this.

**Comment by bryanculver** [comment]
> OK I saw
> 
> > (1) introduce the concept of buildings such that one or more buildings can be added per site
> > (2) consider creating a mechanism (using labels or some other means) to change the name of these organization objects.
> 
> Which to me was the Location model but @chadell's update constrains that. Since it doesn't look like we have an issue already open for a generic contact model (I could have swore we did) I can update this.

We did: https://github.com/nautobot/nautobot/issues/1066 which seemed to get moved to a discussion https://github.com/nautobot/nautobot/discussions/1071

I'll get all of these feedback collapsed.

**Comment by chadell** [comment]
I want to add an extra use case for this Feature Request.

Assuming we have the `Contact` and `Team` models available, I think that a common use case will be to **associate** either a `Contacts` or `Teams` to any type of model with an associated `Role`. So this **association** (`Team` + `Role` or `Contact` + `Role`) will be connected to one or many object (of any type).

For example, having:

```
Contacts:
  -  ContactA
  -  ContactB
 
Teams:
  - Team1
  - Team2

Role:
  - Admin
  - Viewer
  - Operator
```

I may want to associate (I use `RoleAssociation` as an example) them with VRFs:

- VRF Orange
  - RoleAssociations:
    - Team1 + Admin
    - ContactA + Viewer
- VRF Blue
  - RoleAssociations :
    - Team1 + Admin (this is the previous association reused)
    - ContactA + Operator (this is a different association from the previous ContactA)
    - ContactB + Viewer (many contacts and teams can be associated, but the same contact and team can't be duplicated in the same RoleAssociation)

The goal would be to make this `RoleAssociation` generic enough that another use case could be to define the roles of `Contacts` and `Teams` to a `Location`.

**Comment by netgitdom** [comment]
I want to specify @chadell extra use case with the proposed `RoleAssociation` one step further.
We need to assign a `Contact` or a `Team` to multiple `Roles` for one VRF (and on a Parent `Prefix` level as well)

So this keeps the same but with more Roles:
```
Contacts:
  -  ContactA
  -  ContactB
  -  ContactC
 
Teams:
  - Team1
  - Team2

Role:
  - Admin
  - Viewer
  - Operator
  - Security
  - Manager

```

* VRF Orange
  * RoleAssociations:
    * Team1 + Admin
    * ContactA + Viewer
    * ContactA + Security
  * VRF Orange Parent Prefix
    * Roles gets inherited from VRF Level
    * VRF Orange Child Prefix Level (Parent got divided in several UseCases in our case)
      * ContactC + Admin ( for this Child Prefix UseCase is another Admin the contact person )

* VRF Blue
  *RoleAssociations :
    * Team1 + Admin (this is the previous association reused)
    * ContactA + Operator (this is a different association from the previous ContactA)
    * ContactB + Admin (many contacts and teams can be associated, but the same contact and team can't be duplicated in the same RoleAssociation)
    * ContactB + Manager


This is necessary as in our company one person or team can has multiple `Roles` which is used a lot.

**Comment by glennmatthews** [comment]
- Models:
	- Contact (an individual person, has name/email/phone/etc.)
	- Team (a group of persons, has name/email/phone/etc. and may contain a set of individual Contacts)
		- Not a tree model at this time (no parent/child teams)
	- Role (existing generic data model)
	- ContactAssociation (through table between a Contact or a Team and any other Nautobot object)
		- ForeignKey to a Contact xor a Team (two mutually exclusive foreign keys)
		- GenericForeignKey to an arbitrary object instance
			- Add a ReverseRelation to supported models (probably a mixin) so deletes will cascade
		- ForeignKey to a Role (optional/nullable?)
		- ForeignKey to Status (required)
		- Support for Custom Fields
		- Uniqueness constraints are (contact, object, role) and (team, object, role)
			- Need to take nullability and NULL != NULL into account
- UI/UX:
	- Contacts tab on detail view of all supported models (related objects table pattern, again)
		- “Assign a contact” button and appropriate form(s)
	- Button on Location detail view to create and associate a Contact based on its current contact fields
	- Management command to create Contacts from all existing Location contact fields?
- Future (out of current scope):
	- Team as tree model (if there’s a clear need; Nautobot shouldn’t be the SoT for org structure)
	- Inheritance of contacts from other models (part of a broader data inheritance feature epic)
		- Namespace contacts -> Prefixes in that namespace if not overridden
		- Location contacts -> child Locations if not overridden
		- Etc.
	- Future major release (3.0) will remove the current Location contact fields

**Comment by itdependsnetworks** [comment]
Just to be clear, we would not use the "User" model? If so, would be good to document why for future understanding.

**Comment by glennmatthews** [comment]
Correct. Not all Users are Contacts and not all Contacts are Users - for example, a vendor or provider might be a Contact but not a User, and a service account might be a User but not a Contact.
